### PR TITLE
🎉 add interactive globe / TAS-940

### PR DESCRIPTION
**Pull request by sophiamersmann**
---
⧓ Review in [Butler Review `#9065aggVu`](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu)

**🎉 add interactive globe**

16 commit series (version 61)

| Series | Commit Title | Status | Reviewers | 
| --- | --- | --- | --- |
| 16/16 | [🔨 fix rebase](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/81575103-eade-4376-b224-10f3d3c08d69) | ⏳ |  |
| 15/16 | [🔨 minor refactor](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/ded29674-d268-46e1-97db-c1dc525f4849) | ⏳ |  |
| 14/16 | [✨ add panning strategy for cursor outside of the globe](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/02231f0d-ca59-4f08-afdf-f26cb0d7db5f) | ⏳ |  |
| 13/16 | [✨ improve globe panning](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/60e11bd4-0ed1-4154-a5cc-e89f870e4e2d) | ⏳ |  |
| 12/16 | [🐛 fix tooltip disappearing too quickly on touch devices](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/0f30dbef-c71c-4a55-b418-ebefe4fdda4d) | ⏳ |  |
| 11/16 | [🔨 refactor event handling](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/199e7a89-c3ae-46c1-ad80-64489fedf2e1) | ⏳ |  |
| 10/16 | [✨ pin tooltip to bottom for all touch devices](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/41bfa3ab-d551-4a44-a664-79d6ae9f9ab7) | ⏳ |  |
| 9/16 | [🐛 show tooltip on click for touch devices](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/5ab635fb-48e0-4fc8-8918-1315f4d2107e) | ⏳ |  |
| 8/16 | [🐛 slightly increase max height of tooltip](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/44304393-2554-4611-871a-31b55256831c) | ⏳ |  |
| 7/16 | [🐛 remove outline when tooltip is dismissed](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/8339ea4f-b278-4d5a-80da-5102dca3d0d0) | ⏳ |  |
| 6/16 | [✨ update graticule colour](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/a8bd4e8b-96b4-49ae-b58f-f7ad05092f9f) | ⏳ |  |
| 5/16 | [🐛 clip globe to the map chart bounds](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/dd73fcf9-3c1d-40ab-95f6-a2af831a081e) | ⏳ |  |
| 4/16 | [🐛 fix the pinch-to-zoom gesture](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/478b1795-66b4-4fce-8982-cfe90b90373f) | ⏳ |  |
| 3/16 | [🐛 combine panning and zooming](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/4210130d-b71a-407b-8c09-1171fdeaaee1) | ⏳ |  |
| 2/16 | [🔨 use d3 for dragging and zooming](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/4d346f55-1795-4b0c-99d0-26d3e44c9f2b) | ⏳ |  |
| 1/16 | [🎉 add globe](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu/commit/63fcbb65-20fc-407c-a647-aec1b887d1c9) | ⏳ |  |

_Please leave review feedback in the [Butler Review](https://gitbutler.com/sophiamersmann/owid-grapher/reviews/9065aggVu)_

### CODE REVIEW

File: packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx [hunk 1]
```diff
@@ -0,0 +1,622 @@
...
+                    }
+
+                    const delta = distance - startDistance
+
+                    // We sometimes get two events for the same pinch gesture,
+                    // with one of the events having a delta of 0. We simply
+                    // ignore the delta-0 events. This fixes a bug where the
+                    // rendered SVG country paths would interfere with the
+                    // pinch-to-zoom gesture.
+                    if (delta === 0) return
```
Comment (marcelgerber) [inline packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx:418]: Thinking about this once more:
While this line here can obviously stay in, because we don't need to do anything about deltas of 0, in my testing I've also seen a bunch of times where we get two pinch events that have non-zero deltas (but where the second one is miniscule).
Which makes me think that we should make `zoomGlobe` more robust, which would mean that it can properly handle two incoming zoom events before the `requestAnimationFrame` has finished (by somehow storing the "target zoom level").
I think this will be beneficial, and will make it less likely that we behave weirdly in other browsers, or on low-perf devices, etc.

But only if that's relatively easy to do.

File: packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx [hunk 2]
```diff
@@ -0,0 +1,622 @@
...
+        return this.manager.isStatic
+            ? this.renderStatic()
+            : this.renderInteractive()
+    }
+}
+
+const isMultiTouchEvent = (
+    event: MouseEvent | TouchEvent
+): event is TouchEvent => {
+    return checkIsTouchEvent(event) && event.touches.length >= 2
```
Comment (marcelgerber) [inline packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx:613]: We should probably only make it a pinch if it's _exactly two_ touches?

**Review by marcelgerber — COMMENTED** [review]
Really nice work here, this will be a much-appreciated feature!

File: packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx [hunk 3]
```diff
@@ -0,0 +1,622 @@
...
+                previousType = type
+            }
+
+            const panningOrZooming = (event: any): void => {
+                this.isPanningOrZooming = true
+                this.clearHover() // dismiss the tooltip
+
+                const wheeling = (): void => {
+                    this.zoomGlobe(-event.sourceEvent.deltaY)
+                }
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx:401]: Yes, that would definitely be nicer! When I implemented it, there was some jitter—apparently caused by D3—that I couldn’t easily get rid of, so I dropped it. I’m not sure exactly which version of the code that was, so I could give it another shot and see if I can get it working now.

File: packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx [hunk 4]
```diff
@@ -0,0 +1,622 @@
...
+                    // ignore the delta-0 events. This fixes a bug where the
+                    // rendered SVG country paths would interfere with the
+                    // pinch-to-zoom gesture.
+                    if (delta === 0) return
+
+                    this.zoomGlobe(delta)
+                    startDistance = distance
+                }
+
+                const panning = (): void => {
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx:424]: yeah, that's a bit annoying. maybe it would be easiest/best to switch to the naive approach (simply adding x,y) when leaving the globe

File: packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx (hunk 2 above)
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx:613]: I don't think it matters much, but I can change it

File: packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx:424]: Seems to work!

File: packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx [hunk 5]
```diff
@@ -0,0 +1,711 @@
...
+                .on("start", start)
+                .on("zoom", panningOrZooming)
+                .on("end", end)
+        }
+
+        select(base).call(panAndZoom())
+    }
+
+    componentDidMount(): void {
+        document.addEventListener("touchstart", this.onDocumentClick, true)
```
Comment (marcelgerber) [inline packages/@ourworldindata/grapher/src/mapCharts/ChoroplethGlobe.tsx:541]: We should make this into a passive event listener, so it doesn't interfere with scrolling or any other touch interactions:
```suggestion
        document.addEventListener("touchstart", this.onDocumentClick, { capture: true, passive: true })
```

**Comment by sophiamersmann** [comment]
@marcelgerber I pushed a new version where rotating the globe works even if your cursor has left the globe.

I tried making rotate-to-the-zoom-center work but there is still a bit of a jitter that I'm not sure how to get rid of and I also struggled to make it work for pinching on mobile. I'm inclined to park this as a nice-to-have for now.

**Comment by marcelgerber** [comment]
yeah, makes sense to me. thank you!
